core/src/main/java/org/pentaho/di/connections/vfs/BaseVFSConnectionProvider.java

```diff
@@ -90,4 +101,65 @@ protected static boolean getBooleanValueOfVariable( VariableSpace space, String
   protected VariableSpace getSpace( ConnectionDetails connectionDetails ) {
     return connectionDetails.getSpace() == null ? Variables.getADefaultVariableSpace() : connectionDetails.getSpace();
   }
+
+  @Override
+  public boolean test( T connectionDetails, VFSConnectionTestOptions connectionTestOptions ) throws KettleException {
+    boolean valid = test( connectionDetails );
+    if ( !valid ) {
+      return false;
+    }
+
+    if ( !connectionDetails.isSupportsRootPath()  || connectionTestOptions.isIgnoreRootPath() ) {
+      return true;
+    }
+
+    String resolvedRootPath = getResolvedRootPath( connectionDetails );
+    if ( StringUtils.isEmpty( resolvedRootPath ) ) {
+      return !connectionDetails.isRootPathRequired();
+    }
+
+    String internalUrl = buildUrl( connectionDetails, resolvedRootPath );
+    FileObject fileObject = KettleVFS.getFileObject( internalUrl, new Variables(), getOpts( connectionDetails ) );
+
+    try {
+      return fileObject.exists();
+    } catch ( FileSystemException fileSystemException ) {
+      LOGGER.error( fileSystemException.getMessage() );
+      return false;
+    }
+  }
+
+  @Override
+  public String getResolvedRootPath( T connectionDetails ) {
+    if ( StringUtils.isNotEmpty( connectionDetails.getRootPath() ) ) {
+      VariableSpace space = getSpace( connectionDetails );
+      String resolvedRootPath = getVar( connectionDetails.getRootPath(), space );
+      if ( StringUtils.isNotBlank( resolvedRootPath ) ) {
+        return normalizeRootPath( resolvedRootPath );
+      }
+    }
+
+    return StringUtils.EMPTY;
+  }
+
+  private String normalizeRootPath( String rootPath ) {
```

Comment:
A few questions/comments. I might be ignorant on what was decided, we can continue this offline, but I'll pose my them her for now:

1. Is the implementation for `BaseVFSConnectionProvider#normalizeRootPath` and  `BaseVFSConnectionProvider#buildUrl` are a stop gap until the POC dev branches from https://hv-eng.atlassian.net/browse/BACKLOG-40451 are implemented by https://hv-eng.atlassian.net/browse/BACKLOG-40356 ?
2. How does `BaseVFSConnectionProvider#normalizeRootPath` and  `BaseVFSConnectionProvider#buildUrl` related to https://hv-eng.atlassian.net/browse/BACKLOG-40451 ?
3. the logic for ConnectionDetails -> PVFS URI is no this straightforward see ConnectionFileSystem#getUrl - https://github.com/pentaho/pentaho-kettle/blob/f4363fec6f52a94e8506c5614d5541706c5dda7a/core/src/main/java/org/pentaho/di/connections/vfs/provider/ConnectionFileSystem.java#L63-L83. But for root Path testing might this simple
4. if we do go with this code as is for `BaseVFSConnectionProvider#normalizeRootPath` and  `BaseVFSConnectionProvider#buildUrl`, I wouldn't do manual string URI manipulation you an use java's URI construction, it's basically a builder - https://docs.oracle.com/javase/8/docs/api/java/net/URI.html . Pentaho should get out of the business to manually creating URI/URLS.

more references for URI building

- https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#URI-java.lang.String-java.lang.String-java.lang.String-java.lang.String-
- https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#resolve-java.lang.String-
- https://www.baeldung.com/java-url

Reply:
Regarding  1 and 2. The idea is to later replace the code in `buildUrl` with that being developed for https://hv-eng.atlassian.net/browse/BACKLOG-40451, when it is ready.

The `normalizeRootPath` may or may not be needed later, depending on how the new code is written. A normalized root path should start with a `/`, just like `FileName#getPath()` does, in the Apache library. Also, in generic paths, we have been consistently removing the terminating slash, when present. Also, some places assume that a normalized root path can be null, but not empty.

Regarding 3, the logic has been simplified. The SMB condition does not affect PUC atm, given it is not exposed in the PUC UI. The last if, where `hasBuckets` is tested is a bug of the original code, AFAIK, and so was removed. In any case, this is all temporary.

Regarding 4, this is temporary code, essentially, copied from the `getUrl` method you quoted, and there seems to be no advantage and improving it at this point.

Reply:
if it's just temporary, then no sense in putting much effort in it.

Reply:
req have been updated see slack: https://hitachivantara-eng.slack.com/archives/C01HJ874L8Z/p1715101993758869

Reply:
addressed earlier.